### ISSUE TITLE
Question about how to use Recyclable memory stream manager

### ISSUE THREAD

**Issue by Sicos1977** [microsoft/Microsoft.IO.RecyclableMemoryStream#257]
I'm using the memory stream manager in a C# .NET 4.8 project.

I declared the manager in a static class like this

```c#
/// <summary>
///     Stream helper methods
/// </summary>
internal static class StreamHelpers
{
    #region Consts
    private const int BlockSize = 1024;
    private const int LargeBufferMultiple = 1024 * 1024;
    private const int MaxBufferSize = 16 * LargeBufferMultiple;
    #endregion

    #region Fields
    private static RecyclableMemoryStreamManager _manager;
    #endregion

    #region Properties
    /// <summary>
    ///     https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream
    /// </summary>
    internal static RecyclableMemoryStreamManager Manager
    {
        get
        {
            if (_manager != null)
                return _manager;

#if (DEBUG)
            if (_manager != null) 
                _manager.GenerateCallStacks = true;
#endif

            _manager = new RecyclableMemoryStreamManager(BlockSize, LargeBufferMultiple, MaxBufferSize)
            {
                //GenerateCallStacks = true,
                AggressiveBufferReturn = true,
                MaximumFreeLargePoolBytes = MaxBufferSize * 4,
                MaximumFreeSmallPoolBytes = 100 * BlockSize,
            };

            return _manager;
        }
    }
    #endregion
}
```

And use it from code like this:

```c#
                using (var fileData = attachment.FileData)
                using (var memoryStream = StreamHelpers.Manager.GetStream("Converter.ConvertWithConcatenate"))
                {
                    fileData.CopyTo(memoryStream);
                    inputFiles.Add(
                        new TransferFile
                        {
                            Name = attachment.FileName,
                            Data = memoryStream.ToArray(),
                            ChildType = firstAttachment
                                ? ConversionOptionChildtype.FirstChild
                                : ConversionOptionChildtype.Default,
                            PageRange = pageRange
                        });

                    firstAttachment = false;
                }
```

I implemented dispose patterns everywhere to free the streams as soon as they are not needed anymore.

```c#
    public void Dispose()
    {
        if (_disposed) return;

        _attachments?.Dispose();
        _converted?.Dispose();
        _failedFiles?.Dispose();
        _mimeMessage?.Dispose();

        _disposed = true;
    }
```

When I run the code from Visual Studio I see the memory usage go up and down after Garbage collection.

But when I use the code on a production environment I only see the memory usage go up and almost never go down. It can go up to using 8GB of memory and keep allocating that even in the middle of the night when almost nothing is happening on the system.

Am I doing something wrong? Implementing it wrong?

Any advice would be nice,

**Comment by benmwatson** [comment]
The call to `memoryStream.ToArray()` is undoing any possible optimizations of using a pooled stream. The memory is probably going up in production because you have so much spare memory that no GC happens. 

In your case, you would need to keep the stream alive (undisposed) until after you were done with the `TransferFile` it was being used in. If `TransferFile` can only accept a buffer, it should also accept a length so that it knows how much of the buffer is valid.

**Comment by Sicos1977** [comment]
> The call to `memoryStream.ToArray()` is undoing any possible optimizations of using a pooled stream. The memory is probably going up in production because you have so much spare memory that no GC happens.
> 
> In your case, you would need to keep the stream alive (undisposed) until after you were done with the `TransferFile` it was being used in. If `TransferFile` can only accept a buffer, it should also accept a length so that it knows how much of the buffer is valid.

I can't avoid using the ToArray() because it is calling a webservice that only accepts a byte array. Is it in this case better to use a normal MemoryStream instead of a Recyclable one?

**Comment by benmwatson** [comment]
If you can't change the interface to support a length parameter, then there's probably no benefit to using this library, so in that case, I would just stick with regular memorystream.

Ben

**Comment by Sicos1977** [comment]
This goes a little bit above my head how all this works, the reason I started with the RecyclableMemoryStream was because I move a lott of data back and forward. I used to have regular out of memory exceptions even on byte arrays smaller than 200MB. These problems were gone when starting to use this library.

Is it possible to use GetBuffer() instead of ToArray() to get the underlying byte array to avoid that a new copy is made in memory?
